Cython/Build/Tests/TestRecythonize.py

```diff
@@ -1,212 +1,218 @@
-import shutil
 import os
+import shutil
 import tempfile
 import time
 
 import Cython.Build.Dependencies
 import Cython.Utils
 from Cython.TestUtils import CythonTest
 
+SAME = "The result of cytonization is the same"
+INCORRECT = "Incorrect cythonization"
+LINE_BEFORE_IMPLEMENTATION = '  /* "{filename}":{at_line}\n'
+VARS_LINE = '  /*--- Wrapped vars code ---*/\n'
 
-def fresh_cythonize(*args, **kwargs):
-    Cython.Utils.clear_function_caches()
-    Cython.Build.Dependencies._dep_tree = None  # discard method caches
-    Cython.Build.Dependencies.cythonize(*args, **kwargs)
 
 class TestRecythonize(CythonTest):
+    language_level = 3
+    dep_tree = Cython.Build.Dependencies.create_dependency_tree()
+
+    def clear_function_and_Dependencies_caches(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Build.Dependencies._dep_tree = None  # discard method caches
 
     def setUp(self):
         CythonTest.setUp(self)
+        self.clear_function_and_Dependencies_caches()
         self.temp_dir = (
             tempfile.mkdtemp(
                 prefix='recythonize-test',
                 dir='TEST_TMP' if os.path.isdir('TEST_TMP') else None
             )
         )
+        self.src_dir = os.path.join(self.temp_dir, 'src')
+        os.mkdir(self.src_dir)  # self.temp_dir is already a temp
 
     def tearDown(self):
         CythonTest.tearDown(self)
+        self.clear_function_and_Dependencies_caches()
         shutil.rmtree(self.temp_dir)
 
-    def test_recythonize_pyx_on_pxd_change(self):
+    def write_to_file(self, path, text):
+        try:
+            timestamp_before_change = os.path.getmtime(path)
+        except OSError:  # not FileNotFoundError for compatibility
+            timestamp_before_change = .0
+
+        with open(path, "w") as f:
+            f.write(text)
+
+        # Make sure the file has a newer timestamp,
+        # otherwise cythonize does not work as expected
+        # on Linux-like systems #4245
+        while 1:
+            if os.path.getmtime(path) != timestamp_before_change:
+                return
```

Comment:
I could write it like this
```python
while os.path.getmtime(path) == timestamp_before_change:
    pass
```
but I think the current way is easier to understand.

Reply:
What you've written here is an infinite loop. You're not updating the timestamp inside of the loop.
Also, this isn't _always_ necessary, only in a few cases. Maybe we should have a separate method for this to make it explicit in the test functions exactly _when_ this is needed.
Please also add a little `time.sleep(0.01)` to avoid busy looping.
And I very much prefer to have the condition in the where-loop clause, where it belongs.

Reply:
> What you've written here is an infinite loop. You're not updating the timestamp inside of the loop.

Um in the code before that I open the file, so the OS has to update the timestamp ...

Reply:
> And I very much prefer to have the condition in the where-loop clause, where it belongs.

Wdym? Do you prefer `while os.path.getmtime (path) == timestamp_before_change` or what?

Reply:
> Also, this isn't _always_ necessary, only in a few cases. Maybe we should have a separate method for this to make it explicit in the test functions exactly _when_ this is needed.
> Please also add a little `time.sleep(0.01)` to avoid busy looping.

[Done](https://github.com/cython/cython/pull/4241/commits/cdec3726ec8632c2a68823d1465e71c73ea79430), by the way, on average, getting the timestamp update takes literally a few iterations, so reducing the time will make the test faster .. although it does not affect much.

Reply:
`write_to_file_with_timestamp_update` is still not used because I am trying to find the right places to use it

Reply:
Ok, maybe we have different ideas about why this issue arises. My understanding is as follows.

In some tests, we create two files, and the second one is intended to be newer than the first. If we create the files fast enough, they may end up with the same timestamp. This makes the test fail, because `cythonize` considers the relevant file "not newer" and thus "already up to date", and thus refuses to recreate it.

Thus my proposal to _make sure_ that the second file has a newer timestamp than the first that we created.
I intended the new method to be used whenever we create files in the test that need to have different timestamps. I asked for a second method because that makes it explicit in the test code whenever we want a file to be newer. Thus, it should both fix the test and help with understanding them for humans.

Does this match your understanding, or could you explain your view on the matter?

Reply:
> If we create the files fast enough, they may end up with the same timestamp.

I didn't think about it that way. But doesn't my solution imply that the timestamp will be different?
If the file is not created, then there are no problems, and if it has already been created, then I save its timestamp and wait for a new one. Because nowhere were there any problems with infinite loops, the timestamp is always gets an update along the way.

Reply:
> doesn't my solution imply that the timestamp will be different?

Not if my assumptions about the reason for the issue above are correct, no. Unix behaviour isn't random. It will not sit there and think "Oh, should I tell the user that I updated the timestamp? Maybe I'll tell them a little later and finish my cup of tee first." That's not how operating systems work. At least not the ones that people commonly use.

Reply:
As the saying goes, some things are easier done than said. I implemented the helper functions in https://github.com/cython/cython/commit/f7eb64034890495654e0728f0362c848348c6674 – let's see if they do the trick.

Reply:
> I implemented the helper functions in [f7eb640](https://github.com/cython/cython/commit/f7eb64034890495654e0728f0362c848348c6674) – let's see if they do the trick.

I think I can resolve the conversation.